### 문제1: 152쪽 문제 4.9

참게 자료에 대해 색깔변수만을 예측변수로 사용하여 부수체의 확률에 대한 로지스틱 회귀모형을 적합하라.

In [7]:
# crab은 연번
# sat는 부수체의 수
# weight는 무게
# width는 등딱지너비
# color는 색깔(1=약간밝은색, 2=중간색, 3=약간어두운색, 4=어두운색)
# spine은 등뼈의 상태(1=둘다양호함, 2=1개가 휘거나 좋지않음, 3=둘다 휘거나 좋지않음)
# y값도 이미 입력되어 있다.

crab_data = read.table("Crabs.dat.txt", header=TRUE)
head(crab_data)

crab sat y weight width color spine
1 1    8   1 3.05   28.3  2     3    
2 2    0   0 1.55   22.5  3     3    
3 3    9   1 2.30   26.0  1     1    
4 4    0   0 2.10   24.8  3     3    
5 5    4   1 2.60   26.0  3     3    
6 6    0   0 2.10   23.8  2     3

#### a. 색깔변수를 명목형 변수(질적 변수)로 간주하라. 예측식을 구하고 이 식을 이용하여 첫 번째와 네 번째 색깔을 어떻게 비교할 수 있는지 설명하라.

In [8]:
# color 변수를 명목형 변수로 변환
crab_data$colorf = as.factor(crab_data$color)

# 로지스틱 회귀 모델 추정
model1 = glm(y~colorf, family=binomial, data=crab_data)
summary(model1)


Call:
glm(formula = y ~ colorf, family = binomial, data = crab_data)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)  
(Intercept)   1.0986     0.6667   1.648   0.0994 .
colorf2      -0.1226     0.7053  -0.174   0.8620  
colorf3      -0.7309     0.7338  -0.996   0.3192  
colorf4      -1.8608     0.8087  -2.301   0.0214 *
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 225.76  on 172  degrees of freedom
Residual deviance: 212.06  on 169  degrees of freedom
AIC: 220.06

Number of Fisher Scoring iterations: 4


예측식은 다음과 같이 나타낼 수 있다.

<img src="https://latex.codecogs.com/svg.image?logit[P(Y=1)]=log(\frac{\pi}{1-\pi}) =\alpha +\beta_1\times color2+\beta_2\times color3+\beta_3\times color4">

여기서 <img src="https://latex.codecogs.com/svg.image?\pi=P(Y=1)">

적합된 로지스틱 회귀모델을 적용하면 회귀식은 다음과 같다.  
<img src="https://latex.codecogs.com/svg.image?log(\frac{\pi}{1-\pi})=1.0986-0.1226\times color2-0.7309\times color3-1.8608\times color4">

이를 오즈 기준으로 정리하면  
<img src="https://latex.codecogs.com/svg.image?\frac{\pi}{1-\pi}=e^{1.0986-0.1226\times color2-0.7309\times color3-1.8608\times color4}"> 이고,

이를, <img src="https://latex.codecogs.com/svg.image?\pi">기준으로 정리하면   
<img src="https://latex.codecogs.com/svg.image?\pi=\frac{e^{1.0986-0.1226\times color2-0.7309\times color3-1.8608\times color4}}{1+e^{1.0986-0.1226\times color2-0.7309\times color3-1.8608\times color4}}"> 이다.

참게의 색깔이 첫번째 색깔인 경우에는 color2=color3=color4=0 이다.

이 때 부수체가 존재할 오즈는  
<img src="https://latex.codecogs.com/svg.image?\frac{\pi}{1-\pi}=e^{1.0986}\approx 3.000"> 이다.

부수체가 존재할 확률은
<img src="https://latex.codecogs.com/svg.image?\pi=\frac{3}{1+3}\approx 0.75"> 이다.

참게의 색깔이 네번째 색깔인 경우에는 color2=color3=0, color4=1 이다.

이 때 부수체가 존재할 오즈는  
<img src="https://latex.codecogs.com/svg.image?\frac{\pi}{1-\pi}=e^{1.0986-1.8608}\approx 0.467"> 이다.

부수체가 존재할 확률은
<img src="https://latex.codecogs.com/svg.image?\pi=\frac{0.467}{1+0.467}\approx 0.318"> 이다.

#### b. 모형 (a)에서 색깔변수가 효과가 있는지에 대해 가능도비 검정을 하고 결과를 해석하라.

In [9]:
# 단순 모델 (색깔 변수 제외)
model0 <- glm(y ~ 1, data=crab_data, family=binomial)

# 가능도 계산
L_1 <- logLik(model1)
L_0 <- logLik(model0)

# 가능도비 검정 통계량 계산
llh = -2 * (L_0 - L_1)

# 자유도
df <- length(coef(model1)) - 1  # 전체 모델의 계수 수에서 1을 뺀 값

# p-value 계산
p_value <- pchisq(llh, df, lower.tail=FALSE)

# 결과 출력
cat("가능도비 검정 통계량:", llh, "\n")
cat("자유도:", df, "\n")
cat("p-value:", p_value, "\n")


가능도비 검정 통계량: 13.69768 
자유도: 3 
p-value: 0.003346919 


[결과 해석]
가능도비검정 통계량이 13.698, 자유도가 3 이며,
이 때의 p-값은 0.0033 으로 0.05보다 작은 값을 가집니다.

즉, 색깔변수가 부수체 존재 확률에 유의미한 영향을 미친다고 볼 수 있습니다.

#### c. 색깔변수를 질적 변수 대신에 양적 변수(점수 1,2,3,4)로 간주하여 예측식을 구하라. 색깔변수의 효과를 해석하고 색깔변수의 효과가 있는지를 검정하라.

In [10]:
# 질적 변수 대신 양적 변수로 간주하여 예측식 구하기
model2 <- glm(y ~ color, data=crab_data, family=binomial)
summary(model2)


Call:
glm(formula = y ~ color, family = binomial, data = crab_data)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)   2.3635     0.5551   4.257 2.07e-05 ***
color        -0.7147     0.2095  -3.412 0.000645 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 225.76  on 172  degrees of freedom
Residual deviance: 213.30  on 171  degrees of freedom
AIC: 217.3

Number of Fisher Scoring iterations: 4


색깔변수를 양적변수로 간주하여 구한 예측식은 다음과 같다.

<img src="https://latex.codecogs.com/svg.image?logit[P(Y=1)]=log(\frac{\pi}{1-\pi})=2.3635 - 0.7147\times color"><br>

<img src="https://latex.codecogs.com/svg.image?\frac{\pi}{1-\pi}=e^{2.3635 - 0.7147\times color}"><br>  

<img src="https://latex.codecogs.com/svg.image?\pi=\frac{e^{2.3635 - 0.7147\times color}}{1+e^{2.3635 - 0.7147\times color}}"><br>

In [11]:
# 색깔이 1일 때
cat("색깔=1 인 경우\n")
cat("오즈:", exp(2.3635-0.7147), "\n")
cat("확률:", exp(2.3635-0.7147) / (1+exp(2.3635-0.7147)), "\n\n")

# 색깔이 2일 때
cat("색깔=2 인 경우\n")
cat("오즈:", exp(2.3635-0.7147*2), "\n")
cat("확률:", exp(2.3635-0.7147*2) / (1+exp(2.3635-0.7147*2)), "\n\n")

# 색깔이 3일 때
cat("색깔=3 인 경우\n")
cat("오즈:", exp(2.3635-0.7147*3), "\n")
cat("확률:", exp(2.3635-0.7147*3) / (1+exp(2.3635-0.7147*3)), "\n\n")

# 색깔이 4일 때
cat("색깔=4 인 경우\n")
cat("오즈:", exp(2.3635-0.7147*4), "\n")
cat("확률:", exp(2.3635-0.7147*4) / (1+exp(2.3635-0.7147*4)), "\n")

색깔=1 인 경우
오즈: 5.200735 
확률: 0.8387288 

색깔=2 인 경우
오즈: 2.544922 
확률: 0.7179063 

색깔=3 인 경우
오즈: 1.245329 
확률: 0.554631 

색깔=4 인 경우
오즈: 0.6093881 
확률: 0.3786458 


색깔의 값이 1 증가할 때마다 부수체를 가질 오즈와 확률이 감소한다.

그리고 색깔변수에 대한 p값이 0.000645로 매우 작은 값이므로 색깔변수의 값이 부수체를 가지는지 여부에 영향을 준다고 볼 수 있다.

#### d. 색깔변수를 질적 변수 대신에 양적 변수로 다룰 때 검정력에서의 장점과 모형의 적합결여와 관련되어 잠재적인 단점에 대해 논하라.

질적 변수 대신에 양적 변수로 다룰 때 검정력에서의 장점

1. 양적 변수는 더 많은 정보를 표현할 수 있습니다.
- 예를 들어, 질적 변수를 사용할 경우에는 색깔 1(약간 밝은색)과 색깔 4(어두운 색)의 차이가 색깔 1(약간 밝은색)과 색깔 2(중간색)의 차이가 같습니다.  
  하지만, 양적 변수를 사용할 때에는 색깔 1(약간 밝은색)과 색깔 2(중간색)의 차이보다 색깔 1(약간 밝은색)과 색깔 4(어두운 색)의 차이가 더 큰 것을 표현할 수 있습니다.
2. 양적 변수는 사용하면 더 복잡한 상호작용 효과나 비선형 관계를 모델링할 수 있습니다.
- 색깔 1 ~ 색깔 4의 값을 (1,2,3,4)로 입력하지 않고, (1,2,4,8)로 입력하거나 (0.7, 0.4, 0.2, 0.1)로 입력할 수도 있습니다.  
- 모형 적합 과정에서 다항 회귀나 비선형 회귀를 사용할 수 있어 데이터에 대한 모형 적합도를 높일 수 있습니다.

단점: 모형의 적합결여

1. 색깔이 실제로 질적 특성을 가지고 있다면, 이를 양적으로 처리하는 것은 잘못된 가정이 될 수 있습니다.
- 색깔에 순서가 없는데도 숫자로 표현한다면 실제로는 존재하지 않는 관계가 있는 것처럼 왜곡이 발생할 수 있습니다.
2. 과적합의 위험이 증가합니다.
- 양적 변수를 사용하면 더 복잡한 상호작용 효과나 비선형 관계를 모델링할 수 있지만, 이는 과적합의 위험을 증가시킵니다.
- 주어진 데이터에 대한 예측을 잘 한다고 해서 앞으로 얻게 될 데이터를 잘 예측하는 것은 아니기 때문입니다.
3. 해석의 어려움
- 질적 변수를 사용할 때 각 변수의 수준을 정의하기 때문에 해석이 비교적 명확합니다.  
하지만, 양적 변수를 사용하면 숫자의 차이가 실제로 어떤 의미인지 나타내는데 어려움을 겪을 수 있습니다.

#### e. 무게와 색깔변수를 양적 변수로 간주하여 표준화된 계수를 구하고 결과를 해석하라.

In [12]:
model3 = glm(y ~ weight + color, data=crab_data, family=binomial)
summary(model3)

# 표준화 계수 구하기
# 각 변수의 표준편차 구하기
sd_weight = sd(crab_data$weight)
sd_color = sd(crab_data$color)


cat("weight 계수 표준편차:", sd_weight, "\n")
cat("color 계수 표준편차:", sd_color, "\n")

cat("표준화된 weight 계수:", 1.6531/sd_weight, "\n")
cat("표준화된 color 계수:", -0.5142/sd_color)


Call:
glm(formula = y ~ weight + color, family = binomial, data = crab_data)

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  -2.0316     1.1161  -1.820   0.0687 .  
weight        1.6531     0.3825   4.322 1.55e-05 ***
color        -0.5142     0.2234  -2.302   0.0213 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 225.76  on 172  degrees of freedom
Residual deviance: 190.27  on 170  degrees of freedom
AIC: 196.27

Number of Fisher Scoring iterations: 4


weight 계수 표준편차: 0.5770252 
color 계수 표준편차: 0.8019334 
표준화된 weight 계수: 2.864866 
표준화된 color 계수: -0.6412004

표준화계수에 따른 해석은 다음과 같다.  

weight 변수는 값이 커질수록 부수체를 가질 확률이 높아지고, color 변수는 값이 커질수록 부수체를 가질 확률이 낮아진다.  
절대적인 효과의 차이는 약 4.5배이다.  

### 문제2: 155쪽 문제 4.14

<표 4.8>은 감염을 치료하는 위약과 신약을 비교하기 위해 8개의 센터에서 시행한 임상시험 결과자료이다. 각 센터마다 환자들은 처리그룹에 랜덤하게 배정되었다. 이 책은 웹사이트에 있는 Infection 자료 파일을 이용하여 이 자료를 분석하라. 로지스틱 회귀분석모형으로 처리효과를 추론하고 결과를 해석하라.

|센터|처리|성공|실패|
|:-:|:-:|:-:|:-:|
|1|신약|11|25|
|1|대조군|10|27|
|2|신약|16|4|
|2|대조군|22|10|
|3|신약|14|5|
|3|대조군|7|12|
|4|신약|2|14|
|4|대조군|1|16|
|5|신약|6|11|
|5|대조군|0|12|
|6|신약|1|10|
|6|대조군|0|10|
|7|신약|1|4|
|7|대조군|1|8|
|8|신약|4|2|
|8|대조군|6|1|

In [13]:
# center는 연번
# treat는 처리(1=신약, 0=대조군)
data = read.table("Infection.dat.txt", header=TRUE)
head(data)

center treat y  n 
1 1      1     11 36
2 1      0     10 37
3 2      1     16 20
4 2      0     22 32
5 3      1     14 19
6 3      0      7 19

In [14]:
# 로지스틱 회귀 모델 생성
model4 <- glm(cbind(y, n-y) ~ treat + factor(center), family = binomial, data = data)

# 결과 출력
summary(model4)



Call:
glm(formula = cbind(y, n - y) ~ treat + factor(center), family = binomial, 
    data = data)

Coefficients:
                Estimate Std. Error z value Pr(>|z|)    
(Intercept)      -1.3220     0.3165  -4.177 2.95e-05 ***
treat             0.7769     0.3067   2.533  0.01130 *  
factor(center)2   2.0554     0.4201   4.893 9.94e-07 ***
factor(center)3   1.1529     0.4246   2.715  0.00662 ** 
factor(center)4  -1.4185     0.6636  -2.138  0.03255 *  
factor(center)5  -0.5199     0.5338  -0.974  0.33007    
factor(center)6  -2.1469     1.0614  -2.023  0.04310 *  
factor(center)7  -0.7977     0.8149  -0.979  0.32764    
factor(center)8   2.2079     0.7195   3.069  0.00215 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 93.5545  on 15  degrees of freedom
Residual deviance:  9.7463  on  7  degrees of freedom
AIC: 66.136

Number of Fisher Scoring iterations: 4


treat: 처리(1=신약, 0=대조군), center: 센터 구분(1~8까지) 라고 할 때 적합된 예측식은 다음과 같다.

<img src="https://latex.codecogs.com/svg.image?logit[P(Y=1)]=log(\frac{\pi}{1-\pi})=\alpha+\beta_1\times treat+\beta_2\times center2+\beta_3\times center3+\beta_4\times center4+\beta_5\times center5+\beta_6\times center6+\beta_7\times center7+\beta_8\times center8">

적합된 결과를 반영하면

<img src="https://latex.codecogs.com/svg.image?logit[P(Y=1)]=log(\frac{\pi}{1-\pi})=-1.3220+0.7769\times treat+2.0554\times center2+1.1529\times center3-1.4185\times center4-0.5199\times center5-2.1469\times center6-0.7977\times center7+2.2079\times center8">

이를 해석하면
[처리] 신약을 투여했을 때가 대조군을 투여했을 때보다 성공할 오즈가 <img src="https://latex.codecogs.com/svg.image?e^{0.7769}=2.17"> 배 더 높다.
[센터] 센터1을 기준으로 할 때 센터2, 센터3, 센터8은 센터1보다 성공 오즈가 더 높고, 센터4, 센터5, 센터6, 센터7은 성공 오즈가 더 낮다.

그리고, 변수별 p-값을 보면 처리, 센터2, 센터3, 센터4, 센터6, 센터8 변수는 성공 여부에 영향을 미친다고 할 수 있다.

